File: rdkit-sys/wrapper/src/fingerprint.cc
#include "rust/cxx.h"
#include <DataStructs/ExplicitBitVect.h>
#include <GraphMol/Fingerprints/Fingerprints.h>
#include <GraphMol/Fingerprints/MACCS.h>
#include <GraphMol/Fingerprints/MorganFingerprints.h>

namespace RDKit {
std::shared_ptr<ExplicitBitVect> rdk_fingerprint_mol(const std::shared_ptr<ROMol> &mol) {
	return std::shared_ptr<ExplicitBitVect>(RDKFingerprintMol(*mol));
}

std::shared_ptr<ExplicitBitVect> pattern_fingerprint_mol(const std::shared_ptr<ROMol> &mol) {
	return std::shared_ptr<ExplicitBitVect>(PatternFingerprintMol(*mol));
}

std::shared_ptr<ExplicitBitVect> morgan_fingerprint_mol(const std::shared_ptr<ROMol> &mol) {
	return std::shared_ptr<ExplicitBitVect>(MorganFingerprints::getFingerprintAsBitVect(*mol, 3, 2048));
}

std::shared_ptr<ExplicitBitVect> copy_explicit_bit_vect(const std::shared_ptr<ExplicitBitVect> &orig) {
	std::shared_ptr<ExplicitBitVect> fingerprint(new ExplicitBitVect(*orig));
	return fingerprint;
}

std::unique_ptr<std::vector<uint64_t>> explicit_bit_vect_to_u64_vec(const std::shared_ptr<ExplicitBitVect> &bitvect) {
	std::vector<uint64_t> bytes;
	bytes.reserve(bitvect->dp_bits->num_blocks());
	boost::to_block_range(*bitvect->dp_bits, (std::back_inserter(bytes)));
	return std::make_unique<std::vector<uint64_t>>(std::move(bytes));
}

std::shared_ptr<ExplicitBitVect> morgan_fingerprint_mol_with_params(const std::shared_ptr<ROMol> &mol,
                                                                    unsigned int radius, unsigned int n_bits) {
	return std::shared_ptr<ExplicitBitVect>(MorganFingerprints::getFingerprintAsBitVect(*mol, radius, n_bits));
}

std::shared_ptr<ExplicitBitVect> rdk_fingerprint_mol_with_params(const std::shared_ptr<ROMol> &mol,
                                                                 unsigned int min_path, unsigned int max_path,
                                                                 unsigned int fp_size) {
	return std::shared_ptr<ExplicitBitVect>(RDKFingerprintMol(*mol, min_path, max_path, fp_size));
}

std::shared_ptr<ExplicitBitVect> pattern_fingerprint_mol_with_params(const std::shared_ptr<ROMol> &mol,
                                                                     unsigned int fp_size) {
	return std::shared_ptr<ExplicitBitVect>(PatternFingerprintMol(*mol, fp_size));
}

std::shared_ptr<ExplicitBitVect> maccs_fingerprint_mol(const std::shared_ptr<ROMol> &mol) {
	return std::shared_ptr<ExplicitBitVect>(MACCSFingerprints::getFingerprintAsBitVect(*mol));
}

unsigned int explicit_bit_vect_num_bits(const std::shared_ptr<ExplicitBitVect> &bitvect) {
	return bitvect->getNumBits();
}
} // namespace RDKit

File: rdkit-sys/src/bridge/fingerprint.rs
#[cxx::bridge(namespace = "RDKit")]
pub mod ffi {
    unsafe extern "C++" {
        include!("wrapper/include/ro_mol.h");
        include!("wrapper/include/fingerprint.h");

        pub type ROMol = crate::ro_mol_ffi::ROMol;
        pub type ExplicitBitVect;
        pub fn rdk_fingerprint_mol(mol: &SharedPtr<ROMol>) -> SharedPtr<ExplicitBitVect>;
        pub fn pattern_fingerprint_mol(mol: &SharedPtr<ROMol>) -> SharedPtr<ExplicitBitVect>;
        pub fn morgan_fingerprint_mol(mol: &SharedPtr<ROMol>) -> SharedPtr<ExplicitBitVect>;

        pub fn copy_explicit_bit_vect(
            fingerprint: &SharedPtr<ExplicitBitVect>,
        ) -> SharedPtr<ExplicitBitVect>;

        pub fn explicit_bit_vect_to_u64_vec(
            bitvect: &SharedPtr<ExplicitBitVect>,
        ) -> UniquePtr<CxxVector<u64>>;

        // Configurable fingerprints
        pub fn morgan_fingerprint_mol_with_params(
            mol: &SharedPtr<ROMol>,
            radius: u32,
            n_bits: u32,
        ) -> SharedPtr<ExplicitBitVect>;
        pub fn rdk_fingerprint_mol_with_params(
            mol: &SharedPtr<ROMol>,
            min_path: u32,
            max_path: u32,
            fp_size: u32,
        ) -> SharedPtr<ExplicitBitVect>;
        pub fn pattern_fingerprint_mol_with_params(
            mol: &SharedPtr<ROMol>,
            fp_size: u32,
        ) -> SharedPtr<ExplicitBitVect>;
        pub fn maccs_fingerprint_mol(mol: &SharedPtr<ROMol>) -> SharedPtr<ExplicitBitVect>;
        pub fn explicit_bit_vect_num_bits(bitvect: &SharedPtr<ExplicitBitVect>) -> u32;
    }
}
